trie/triedb/pathdb/database.go

```diff
@@ -227,45 +232,58 @@ func (db *Database) Commit(root common.Hash, report bool) error {
 	db.lock.Lock()
 	defer db.lock.Unlock()
 
-	// Short circuit if the database is in read only mode.
-	if db.readOnly {
-		return errSnapshotReadOnly
+	// Short circuit if the mutation is not allowed.
+	if err := db.modifyAllowed(); err != nil {
+		return err
 	}
 	return db.tree.cap(root, 0)
 }
 
-// Reset rebuilds the database with the specified state as the base.
-//
-//   - if target state is empty, clear the stored state and all layers on top
-//   - if target state is non-empty, ensure the stored state matches with it
-//     and clear all other layers on top.
-func (db *Database) Reset(root common.Hash) error {
+// Deactivate disables the database and invalidates all available state layers
+// as stale to prevent access to the persistent state, which is in the syncing
+// stage.
+func (db *Database) Deactivate() error {
 	db.lock.Lock()
 	defer db.lock.Unlock()
 
 	// Short circuit if the database is in read only mode.
 	if db.readOnly {
-		return errSnapshotReadOnly
+		return errDatabaseReadOnly
 	}
-	batch := db.diskdb.NewBatch()
-	root = types.TrieRootHash(root)
-	if root == types.EmptyRootHash {
-		// Empty state is requested as the target, nuke out
-		// the root node and leave all others as dangling.
-		rawdb.DeleteAccountTrieNode(batch, nil)
-	} else {
-		// Ensure the requested state is existent before any
-		// action is applied.
-		_, hash := rawdb.ReadAccountTrieNode(db.diskdb, nil)
-		if hash != root {
-			return fmt.Errorf("state is mismatched, local: %x, target: %x", hash, root)
-		}
+	// Prevent duplicated disable operation.
+	if db.waitSync {
+		return nil
```

Comment:
Can we do a double disable? If not, perhaps error here, or at least an error log?